aptoide-installer/src/main/java/cm/aptoide/pt/installer/fetch/FetchDownloader.kt

```diff
@@ -0,0 +1,215 @@
+    packageName: String,
+    versionCode: Long,
+    installationFiles: Set<InstallationFile>
+  ): Flow<Double> {
+    val totalSize = installationFiles.sumOf { it.fileSize }.toDouble()
+    val fetch = Fetch.Impl.getDefaultInstance()
+
+    val destinationDir = File(downloadsPath, packageName).apply {
+      if (!exists()) {
+        mkdirs().let {
+          if (!it) throw IllegalStateException("Can't create download folder: $downloadsPath/$packageName")
+        }
+      }
+    }
+
+    val requests = installationFiles.mapNotNull { file ->
+      File(destinationDir, file.name)
+        .takeUnless { it.checkMd5(file.md5) }
+        ?.let {
+          Request(file.url, it.toUri())
+            .apply {
+              priority = Priority.NORMAL
+              networkType = NetworkType.ALL
+              groupId = packageName.asGroupId()
+              addHeader(VERSION_CODE, versionCode.toString())
+              addHeader(PACKAGE, packageName)
+              addHeader(FILE_TYPE, file.type.toString())
+            }
+            .also {
+              contentHashes.put(it.id, file.md5)
+            }
+        }
+    }
+
+    val downloads = requests.map { it.id }.toMutableSet()
+    val downloadsProgresses = requests.associate { it.id to 0L }.toMutableMap()
+
+    return callbackFlow<Double> {
+      try {
+        trySend(0.0)
+
+        if (downloads.isEmpty()) {
+          trySend(1.0)
+          close()
+        }
+
+        fetch.enqueue(requests) { enqueuedRequests ->
+          enqueuedRequests.find { it.second != Error.NONE }?.let {
+            close(it.second.throwable ?: IllegalStateException("Error enqueuing requests"))
+          }
+        }
+
+        val listener = object : FetchFileDownloadListener {
+          override fun onCompleted(download: Download) {
+            downloads.remove(download.id)
+            if (downloads.isEmpty()) {
+              trySend(1.0)
+              close()
+            }
+          }
+
+          override fun onError(
+            download: Download,
+            error: Error,
+            throwable: Throwable?
+          ) {
+            if (download.status == Status.FAILED) { //No more retries left
+              close(throwable ?: AbortException("Error downloading files"))
+            }
+          }
+
+          override fun onProgress(
+            download: Download,
+            etaInMilliSeconds: Long,
+            downloadedBytesPerSecond: Long
+          ) {
+            downloadsProgresses.replace(download.id, download.downloaded)
+            trySend(downloadsProgresses.values.sum() / totalSize)
+          }
+        }
+
+        fetch.addListener(listener)
+
+        awaitClose {
+          fetch.removeListener(listener)
+        }
+      } catch (e: Throwable) {
+        close(e)
+      }
+    }
+      .distinctUntilChanged()
+      .onCompletion {
+        if (it != null) {
+          if (it is CancellationException) {
+            fetch.cancelGroup(packageName.asGroupId())
```

Comment:
Is the cancel going to keep the download for future restart and the delete completely remove it?

Reply:
The cancel only cancels the download and does not remove any file. The delete removes all the group files, even successfully downloaded ones. The remove deletes only temporary files.

The logic on this part changed a bit and delete is not used any longer.